Adding shearlab.jl submodule

arsenal9971 opened the pull request:
Adding shearlab.jl backend submodule on odl

pep8speaks commented:
Checking new PR...







 - In the file [`odl/contrib/shearlab/shearlab_operator.py`](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py), following are the PEP8 issues :

> [Line 289:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L289): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 286:84](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 286:82](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 286:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 286:55](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 286:19](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 286:17](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L286): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 283:53](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L283): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 282:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L282): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 278:92](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 278:90](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 278:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 278:55](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 278:49](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 278:47](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 278:14](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L278): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 275:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L275): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 272:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L272): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 272:73](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L272): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 272:40](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L272): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 271:84](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 271:82](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 271:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 271:55](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 271:49](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 271:47](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 271:14](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L271): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 269:51](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L269): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 269:34](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L269): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 268:22](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L268): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 268:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L268): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 264:92](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 264:90](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 264:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 264:55](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 264:19](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 264:17](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L264): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 261:53](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L261): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 260:17](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L260): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 260:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L260): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 257:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L257): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 257:35](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L257): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 256:11](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L256): [E271](https://duckduckgo.com/?q=pep8%20E271) multiple spaces after keyword
> [Line 255:51](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L255): [E703](https://duckduckgo.com/?q=pep8%20E703) statement ends with a semicolon
> [Line 246:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L246): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 79 characters)
> [Line 245:37](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L245): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 244:32](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L244): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 243:19](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L243): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 242:26](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L242): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 241:22](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L241): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 240:19](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L240): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 239:19](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L239): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 238:62](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L238): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 238:54](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L238): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 238:33](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L238): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 238:5](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L238): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 238:5](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L238): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 236:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L236): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 236:56](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L236): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 236:33](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L236): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 235:80](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L235): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 235:52](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L235): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 235:50](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L235): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 235:33](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L235): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 234:38](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L234): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 234:33](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L234): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 233:45](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L233): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 233:33](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L233): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 232:34](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L232): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 232:29](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L232): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 217:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L217): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 204:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L204): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 200:60](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L200): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 200:48](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L200): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 200:44](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L200): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 200:35](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L200): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 200:31](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L200): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:61](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:54](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:50](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:46](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:41](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 199:37](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L199): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 198:5](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L198): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 198:5](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L198): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 196:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L196): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 193:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L193): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 191:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L191): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 184:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L184): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 181:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L181): [E266](https://duckduckgo.com/?q=pep8%20E266) too many leading '#' for block comment
> [Line 145:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L145): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 109:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L109): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 27:1](https://github.com/odlgroup/odl/blob/bfbed526a77d450e519be87c4725a8e85b6548e9/odl/contrib/shearlab/shearlab_operator.py#L27): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

 - Complete extra results for this file :

> class ShearlabOperator(odl.Operator):^class ShearlabOperatorInverse(odl.Operator):^class ShearlabOperatorAdjointInverse(odl.Operator):^## Python library for shearlab.jl^def load_Shearlab():^^j = load_Shearlab()^def load_image(name, n, m=None, gpu=0, square=0):^    	m = n    ^    	m = n    ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                    ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                        ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                             ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                                 ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                                     ^    command = 'Shearlab.load_image('+'"'+name+'"'+','+str(n)+','                                                            ^    command = command + str(m)+','+str(gpu)+','+str(square)+')'                              ^    command = command + str(m)+','+str(gpu)+','+str(square)+')'                                  ^    command = command + str(m)+','+str(gpu)+','+str(square)+')'                                           ^    command = command + str(m)+','+str(gpu)+','+str(square)+')'                                               ^    command = command + str(m)+','+str(gpu)+','+str(square)+')'                                                           ^def imageplot(f, str='', sbpt=[]):^class Shearletsystem2D:^def getshearletsystem2D(rows,cols,nScales,                            ^def getshearletsystem2D(rows,cols,nScales,                                 ^                                shearLevels= None,                                ^                                shearLevels= None,                                            ^                                full= 0,                                ^                                full= 0,                                     ^                                directionalFilter = 'Shearlab.filt_gen("directional_shearlet")',                                ^                                directionalFilter = 'Shearlab.filt_gen("directional_shearlet")',                                                 ^                                directionalFilter = 'Shearlab.filt_gen("directional_shearlet")',                                                   ^                                directionalFilter = 'Shearlab.filt_gen("directional_shearlet")',                                                                               ^                                quadratureMirrorFilter= 'Shearlab.filt_gen("scaling_shearlet")'):                                ^                                quadratureMirrorFilter= 'Shearlab.filt_gen("scaling_shearlet")'):                                                       ^                                quadratureMirrorFilter= 'Shearlab.filt_gen("scaling_shearlet")'):                                                                               ^    	shearLevels = [float(ceil(i/2)) for i in range(1,nScales+1)]    ^    	shearLevels = [float(ceil(i/2)) for i in range(1,nScales+1)]    ^    	shearLevels = [float(ceil(i/2)) for i in range(1,nScales+1)]    	                           ^    	shearLevels = [float(ceil(i/2)) for i in range(1,nScales+1)]    	                                                ^    	shearLevels = [float(ceil(i/2)) for i in range(1,nScales+1)]    	                                                        ^    j.eval('rows='+str(rows))                  ^    j.eval('cols='+str(cols))                  ^    j.eval('nScales='+str(nScales))                     ^    j.eval('shearLevels='+str(shearLevels))                         ^    j.eval('full='+str(full))                  ^    j.eval('directionalFilter='+directionalFilter)                               ^    j.eval('quadratureMirrorFilter='+quadratureMirrorFilter)                                    ^    j.eval('shearletSystem = Shearlab.getshearletsystem2D(rows,cols,nScales,shearLevels,full,directionalFilter,quadratureMirrorFilter);')                                                                               ^    isComplex = j.eval('shearletSystem.isComplex');                                                  ^    return  Shearletsystem2D(shearlets, size, shearLevels, full, nShearlets,          ^                                  shearletIdxs, dualFrameWeights, RMS, isComplex)                                  ^                                  shearletIdxs, dualFrameWeights, RMS, isComplex)                                                                               ^def sheardec2D(X,shearletSystem):^def sheardec2D(X,shearletSystem):                ^    coeffs = np.zeros(shearletSystem.shearlets.shape,dtype=np.complex_)                                                    ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                  ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                                                      ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                                                                               ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                                                                                         ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*np.conj(shearletSystem.shearlets[:,:,i]))))                                                                                           ^def shearrec2D(coeffs,shearletSystem):^def shearrec2D(coeffs,shearletSystem):                     ^    X = np.zeros((coeffs.shape[0],coeffs.shape[1]),dtype=np.complex_)                                 ^    X = np.zeros((coeffs.shape[0],coeffs.shape[1]),dtype=np.complex_)                                                  ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]             ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                              ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                                ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                                      ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                                                               ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                                                                 ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*shearletSystem.shearlets[:,:,i]                                                                                   ^    return (fftshift(ifft2(ifftshift((1/shearletSystem.dualFrameWeights)*X)))).real                                       ^    return (fftshift(ifft2(ifftshift((1/shearletSystem.dualFrameWeights)*X)))).real                                                                        ^    return (fftshift(ifft2(ifftshift((1/shearletSystem.dualFrameWeights)*X)))).real                                                                               ^def sheardecadjoint2D(coeffs, shearletSystem):^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])             ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                              ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                                ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                                      ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                                                               ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                                                                         ^        X = X+fftshift(fft2(ifftshift(coeffs[:,:,i])))*np.conj(shearletSystem.shearlets[:,:,i])                                                                                           ^def shearrecadjoint2D(X, shearletSystem):^    coeffs = np.zeros(shearletSystem.shearlets.shape,dtype=np.complex_)                                                    ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                  ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                                                      ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                                                                               ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                                                                                 ^        coeffs[:,:,i] = fftshift(ifft2(ifftshift(Xfreq*shearletSystem.shearlets[:,:,i])))                                                                                   ^^---

arsenal9971 commented:
@adler-j Jonas, here is the pull-request, it seems that the writing style is not matching with the one used by you, should I change this?

adler-j commented:
ODL uses the official python styleguide as in [PEP8](https://www.python.org/dev/peps/pep-0008/), so all code that goes in has to follow this style. So yeah, has to be updated

arsenal9971 commented:
I just did it, I guess now is ready to merge, but the tests does not start again.

arsenal9971 commented:
I think now it works

kohr-h commented:
Very cool to see this coming in! Would you like a review @arsenal9971?

arsenal9971 commented:
Yes please.

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+# Shearlab.jl
+
+This package functionality for integrating ODL with the [julia API of Shearlab](https://github.com/arsenal9971/Shearlab.jl). This package requires different libraries installed, explained in the following.
```
Yes, that is right, I will change that phrase

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+
+## Dependencies and its installation.
+
+- **Julia language**: One can either installed [precompiled packages](https://julialang.org/downloads/) or build from [source](https://github.com/JuliaLang/julia). This package uses the version 0.6 of julia.
```
Ok, I will, is actually working now just on julia 0.5 and julia 0.6

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+# Shearlab.jl
+
+This package functionality for integrating ODL with the [julia API of Shearlab](https://github.com/arsenal9971/Shearlab.jl). This package requires different libraries installed, explained in the following.
```
Ok, I will write julia as Julia

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+
+- **Shearlab.jl**: To install the library in julia 0.6.x one needs to run the command `julia -e 'Pkg.add("Shearlab")`.
+
+- **Pyjulia**: One can install the python API of julia by building from [source](https://github.com/JuliaPy/pyjulia), or using `pip install julia` for Python 2.x and `pip3 install julia` for Python 3.x.
```
Ok

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+
+- **Shearlab.jl**: To install the library in julia 0.6.x one needs to run the command `julia -e 'Pkg.add("Shearlab")`.
+
+- **Pyjulia**: One can install the python API of julia by building from [source](https://github.com/JuliaPy/pyjulia), or using `pip install julia` for Python 2.x and `pip3 install julia` for Python 3.x.
```
I will change this

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+
+# Shearlet Decomposition function
+def sheardec2D(X, shearletSystem):
```
I will change this then.

arsenal9971 replied:
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+        self.shearletIdxs = shearletIdxs
+        self.dualFrameWeights = dualFrameWeights
+        self.RMS = RMS
+        self.isComplex = isComplex
```
Ok, I did classes since that is the style in julia when doing objects. The so called duck typing, this makes the julia code more efficient for the precompiler.

kohr-h replied:
I see. Well, duck typing is also a big topic in Python, but without the precompiler :-)
Not that it makes any difference here, but tuples tend to be more compact and (perhaps) faster. If I would have had to write this code from scratch, I would have used a dictionary. But it's up to you.

arsenal9971 replied:
I would like to keep this format, it does not affect the performance of the module, and it feels more easy to call, I also think this are attributes of a ShearletSystem that should be used in the future in the implemenation

kohr-h replied:
Okay, fair enough. But then please inherit from `object`, for our friends who are stuck in the past (Python 2).

adler-j replied:
Holgers comment here should be adressed

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+# Shearlet Decomposition function
+def sheardec2D(X, shearletSystem):
+    coeffs = np.zeros(shearletSystem.shearlets.shape, dtype=np.complex_)
+    Xfreq = fftshift(fft2(ifftshift(X)))
```
I will disagree on you with the heavy work is done in Python. Actually the heavy work in the Shearlet transform implementation is done in the generation of the corresponding system which in this case is in julia, the decomposition takes nothing in comparison with the generation. Of course there is a implementation of all that function in julia, since my library Shearlab.jl is fully written in julia (check the corresponding [implemenations](https://github.com/arsenal9971/Shearlab.jl/blob/master/src/2D/shearletdecrec2D.jl)).

The reason why I tried to implement this in python is more related on how pyjulia works, the object called Shearlet system is a special type in julia created by me. When trying to import this from julia to python I am not sure how this will work. I also don't know well how to import julia objects into the python enviroment, i know julia arrays can be imported as numpy arrays, I used this fact in the implementation. I think I have a good idea to do this, just give me between today and tomorrow and I will find the way to do also the decomposition and everything in julia. I also will run some benchmarks which case is better.

kohr-h replied:
Okay, fair enough. There you see my ignorance of what happens under the hood :-)
What's your rough estimate on the distribution of the numerical effort between generating the system and performing the decomposition? I'm asking to see if it may still be worth to improve efficiency on the Python side.

arsenal9971 replied:
I runned some [benchmarks](https://github.com/arsenal9971/odl/blob/master/odl/contrib/shearlab/examples/Benchmarks.ipynb), and you can see that the system generation takes most of the time, and is way faster in my version than pyshearlab (around 9 seconds vs 44). You can also see that the decomposition and reconstruction doesnt take so much time (both around 2 seconds), and running pure julia code it just improve around .5 seconds.

 It will also take time if I want to do everything in julia to stream the numpy array that we want to analyse to a julia array, and it will also take more memory, therefore I dont think is worthwhile to do a pure julia implementation of this. I also think that adding a todo to use pyFFTW is a good idea.

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+def shearrec2D(coeffs, shearletSystem):
+    X = np.zeros((coeffs.shape[0], coeffs.shape[1]), dtype=np.complex_)
+    for i in range(shearletSystem.nShearlets):
+        X = X + fftshift(fft2(ifftshift(coeffs[:, :, i]))) *
```
Ok, I will change this

kohr-h replied:
Just to be sure: the second thing is up to you to decide, based on typical memory requirement. If doing a complete FFT requires a too big workspace, this implementation is preferred.

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+# Shearlet Recovery function
+def shearrec2D(coeffs, shearletSystem):
+    X = np.zeros((coeffs.shape[0], coeffs.shape[1]), dtype=np.complex_)
```
Ok, give me some time to come up with a new version

arsenal9971 replied:
Done

adler-j commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+        X = X + fftshift(fft2(ifftshift(coeffs[:, :, i]))) *
+        shearletSystem.shearlets[:, :, i]
+    return (fftshift(ifft2(ifftshift((
+            1 / shearletSystem.dualFrameWeights) * X)))).real
```
why not `X / shearletSystem.dualFrameWeights`, one less operation :D

adler-j commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+def sheardecadjoint2D(coeffs, shearletSystem):
+    X = np.zeros((coeffs.shape[0], coeffs.shape[1]), dtype=complex)
+    for i in range(shearletSystem.nShearlets):
+        X = X + fftshift(fft2(ifftshift(coeffs[:, :, i]))) *
```
Does this even work? It looks like `*` is hanging (one missing parenthesis)

adler-j commented on odl/contrib/shearlab/test/operator_test.py:
```diff
@@ -0,0 +1,57 @@
+from odl.util.testutils import all_almost_equal, simple_fixture
+
+
+dtype = simple_fixture('dtype', ['float32', 'float64'])
```
So here we have an option in dtypes (single/double precision) but in the code it seems like you only support double precision? Is this the case for the Julia implementation as well?

adler-j commented on odl/contrib/shearlab/test/operator_test.py:
```diff
@@ -0,0 +1,57 @@
+
+    # <Ax, Ax> = <x, AtAx>
+    ax = op.adjoint(y)
+    assert pytest.approx(y.inner(y), rel=rel) == phantom.inner(ax)
```
I guess this is my code by these days we try to write this the other way around:

```
assert something == pytest.approx(something_else)
```

since this flows better with english "something is approximately something else"

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+- **Pyjulia**: One can install the python API of julia by building from [source](https://github.com/JuliaPy/pyjulia), or using `pip install julia` for Python 2.x and `pip3 install julia` for Python 3.x.
+   - One also needs to make the julia and python enviroment to coincide running the command `julia -e 'ENV["PYTHON"]="/usr/bin/python"; Pkg.add("PyCall"); Pkg.build("PyCall")'`. Where the directory of the python binary needs to be the binary of the enviroment you want to use.
+
+- **SSL certificates**: Sometimes you need to give (and add to bashrc.) the SSL certificates path using `export SSL_CERT_FILE=/etc/ssl/ca-bundle.pem`.
```
Julia uses the library LibGit2 for its package manager, sometimes if the SSL certificates are not defined in the enviroment, this doesnt work.

kohr-h replied:
This explanation would be perfect for the README :wink:

arsenal9971 commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,14 @@
+# Shearlab.jl
```
Done

arsenal9971 commented on odl/contrib/shearlab/examples/wave_shear_separation.py:
```diff
@@ -0,0 +1,58 @@
+    min_{w,s}  ||W(w)||_1 + a ||S(s)||_1 + b ||w + s - x||_2^2
+
+Here we expect the shearlet coefficients to contain the edges, and the wavelet
+coefficients to contain the square.
```
I copy this from the examples of the submodule of pyshearlab. This should be changed then in the original

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+    if m is None:
+        m = n
+    command = 'Shearlab.load_image(' + '"' + name + '"' + ',' + str(n) + ','
+    command = command + str(m) + ',' + str(gpu) + ',' + str(square) + ')'
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+    @property
+    def adjoint(self):
+        """Return the adjoint operator."""
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+class ShearlabOperator(odl.Operator):
+
+    """Shearlet transform using Shearlab.jl as backend.
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+class ShearlabOperator(odl.Operator):
+
+    """Shearlet transform using Shearlab.jl as backend.
+    This is the non-compact shearlet transform implemented using the fourier
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+class ShearlabOperatorAdjoint(odl.Operator):
+
+    """Adjoint of the shearlet transform.
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+        return ShearlabOperatorInverse(self)
+
+
+class ShearlabOperatorAdjoint(odl.Operator):
```
Done, thanks for the help

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+
+# Function to load Shearlab
+def load_Shearlab():
```
I already changed the name, I dont think there is a difference on giving this outside a function or not. I like to have a function that initiates the libraries, and we need the returned variable to run more julia code, you can see it in the code bellow.

kohr-h replied:
I understand that, I was wondering what the benefits are over doing this:
```python
# Right after import
j = julia.Julia()
j.foo()

# Further down
j.bar()
```
You define a function and then run it immediately to get a `j` instance. So unless you use or plan to use the `load_...` function somewhere else, for instance to run multiple Julia instances for some reason, I don't quite see the benefit.

But anyway, your call.

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+# Function to plot images
+def imageplot(f, str='', sbpt=[]):
+    """
```
It plots figures generated by the library, i just included the summary

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+import julia
+import matplotlib.pyplot as plt
+import pylab
+from math import ceil
```
ok

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+import julia
+import matplotlib.pyplot as plt
+import pylab
```
I just got rid of it

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+import odl
+import numpy as np
+# import shearlab
```
Done

arsenal9971 commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+
+if __name__ == '__main__':
+    # pylint: disable=wrong-import-position
```
Done

arsenal9971 commented:
@adler-j  and @kohr-h I will continue tomorrow on solving the issues given by @adler-j. You can find a [test](https://github.com/arsenal9971/odl/blob/master/odl/contrib/shearlab/examples/Shearlab_odl_test.ipynb) that I did to show the functionality of this submodule. Also notice that some pep8 issues are not so compatible with pyshearlab. Since one need to run strings that can be long and cannot be send to the line bellow without crashing, any suggestions are welcome.

kohr-h commented:
> @adler-j and @kohr-h I will continue tomorrow on solving the issues given by @adler-j.

:+1: 



> You can find a test that I did to show the functionality of this submodule.

:tada: 



> Also notice that some pep8 issues are not so compatible with pyshearlab. Since one need to run strings that can be long and cannot be send to the line bellow without crashing, any suggestions are welcome.

That seems solvable. For instance, you can write 
```python
j.eval('shearletSystem=Shearlab.getshearletsystem2D(rows, cols, nScales, shearLevels, full, directionalFilter, quadratureMirrorFilter) ')
```
as 
```python
j.eval('shearletSystem=Shearlab.getshearletsystem2D(rows, cols, nScales, '
       'shearLevels, full, directionalFilter, quadratureMirrorFilter) ')
```
Python will concatenate the strings without extra newlines. This works with as many lines as you want.

arsenal9971 commented:
@adler-j @kohr-h I just fixed all the issues you mentioned, I guess is now ready to merge. Just to answer @adler-j [question](https://github.com/odlgroup/odl/pull/1322#discussion_r180745034), indeed the Shearlab.jl supports different precisions (actually for the gpu version it supports at most float32). 

I also just modified the adjoint slightly so now it behaves as it supposed to, one can see that in [test](https://github.com/arsenal9971/odl/blob/master/odl/contrib/shearlab/examples/Shearlab_odl_test.ipynb)

kohr-h commented on odl/contrib/shearlab/examples/Benchmarks.ipynb:
```diff
@@ -0,0 +1,526 @@
+{
```
Great to have benchmarks!

Just two remarks on this:
- Please import the functions from the module instead of copying the code into the notebook:
   ```python
   from odl.contrib.shearlab.shearlab_operator import (
       <function names go here>)
   ```
- If you like, use `odl.util.Timer` instead of `tic` and `toc`, the output will be a bit more informative:
   ```python
   >>> with odl.util.Timer('Time for sleep(1.5)'):
   ...     sleep(1.5)
   Time for sleep(1.5) :      1.513
   ```

arsenal9971 commented:
Yes sure

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+
+        """Initialize a new instance.
+        Parameters
+        ----------
```
Sorry for insisting on this. @adler-j meant that there should be a blank line *between summary and "Parameters" heading*, not above the whole string. The current layout confuses the documentation engine Sphinx that is also used in Spyder for rendering the docstrings.

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+            The number of scales for the shearlet transform, higher numbers
+            mean better edge resolution but more computational burden.
+        Examples
+        --------
```
Same here: each of these section headings need a blank line above them, otherwise they're not recognized.

adler-j commented:
Quite busy now so can't do a full review, but overall I'm quite sceptical to adding more figures to the core ODL library. Basically this causes an uncontrolled growth in the repository size that we can never backtrack on. Even small additions like 100kb here and there add up to huge amounts (gigabytes) over several years.

I'd propose we leave the examples in the folder on a "classical level", and then perhaps move the IPython notebooks to e.g. a gists (which can be linked from the repository).

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+        class ShearlabOperatorAdjoint(odl.Operator):
+
+            """Adjoint of the shearlet transform.
+            Should not be used independently.
```
If you replace this line with a blank line, everything is perfect :smiley:

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+
+                class ShearlabOperatorAdjointInverse(odl.Operator):
+
+                    """\
```
Please no backslashes in strings. The summary of this operator is mainly visible to developers, so you can keep it short, like "Inverse adjoint shearlet transform."

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+                    """\
+                    Adjoint of the inverse/Inverse of the adjoint
+                    of shearlet transform.
+                    Should not be used independently.
```
This line can now be removed (everywhere)

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,312 @@
+
+
+# Function to generate de 2D system
+def getshearletsystem2D(rows, cols, nScales, shearLevels=None,
```
What I meant with docstring was something like this:
```python
def getshearletsystem2D(...):
    """Generate a 2D shearlet system."""
```
Basically you can put the comments that you placed above the functions into the docstring.

kohr-h commented:
> Quite busy now so can't do a full review, but overall I'm quite sceptical to adding more figures to the core ODL library. Basically this causes an uncontrolled growth in the repository size that we can never backtrack on. Even small additions like 100kb here and there add up to huge amounts (gigabytes) over several years.
> 
> I'd propose we leave the examples in the folder on a "classical level", and then perhaps move the IPython notebooks to e.g. a gists (which can be linked from the repository).

Forgot to address that, yes. It seems like the image is only used in the benchmarks. So why not just benchmark with zeros and ditch the image?

arsenal9971 commented:
I hope now everything is fine. I just did the fixes you asked. Since I dont see any value on showing the raw code of benchmarks without a notebook, I just took out the benchmarks from the repository and add link in the README of the subdirectory examples to the benchmarks that I add in the Shearlab repository. You can see this [here](https://github.com/arsenal9971/odl/tree/master/odl/contrib/shearlab/examples).

arsenal9971 commented:
I just did a small key change, at the end of the function that creates the shearlet system using julia backend, i delete the very heavy Julia variable once it is already assign to the corresponding python variable, this free up quite a lot of memory. I realized this memory wasn't freed up by the python trash collector. 

Looking forward to merge.

adler-j commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,326 @@
+
+# Function to plot images
+def imageplot(f, str=None, sbpt=None):
+    """
```
Overal som style issues with the comments, this should be

```
def imageplot(f, str=None, sbpt=None):
    """Plot an image generated by the library."""
    pass
```

adler-j commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,326 @@
+j = load_julia_with_Shearlab()
+
+
+# Function to load images with certain size
```
Comment should go inside

kohr-h commented:
> I just did a small key change, at the end of the function that creates the shearlet system using julia backend, i delete the very heavy Julia variable once it is already assign to the corresponding python variable, this free up quite a lot of memory. I realized this memory wasn't freed up by the python trash collector.

Great that you found that out.

---

Before merging we will squash the commits, as described in the `odl.contrib` README. There was a bit of back and forth in the commit history, images and notebooks added and removed again, so we'll clean that up a little bit. You can choose to do that yourself, but it's not a must.

kohr-h commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,15 @@
+# Shearlab.jl
+**Original contribution by:** @arsenal9971  
```
I realized in another PR that this will not automatically generate a link to your GH profile, so you may want to create that using `[@arsenal9971](https://github.com/arsenal9971)` notation

kohr-h commented on odl/contrib/shearlab/README.md:
```diff
@@ -0,0 +1,15 @@
+
+## Dependencies and its installation.
+
+- **Julia language**: One can either installed [precompiled packages](https://julialang.org/downloads/) or build from [source](https://github.com/JuliaLang/julia). This package requires the Julia version 0.6 or higher.
```
install

kohr-h commented:
Hey @arsenal9971, what's the status here? Apart from @adler-j's final remarks, this is good to go. I'd really like to see this merged soon.

arsenal9971 commented:
I already fixed the last issues you pointed out

kohr-h commented on odl/contrib/shearlab/examples/wave_shear_separation.py:
```diff
@@ -0,0 +1,58 @@
+"""Example of using shearlet coefficients for signal separation.
```
How about this, do you want to keep the example @arsenal9971?

kohr-h commented on odl/contrib/shearlab/shearlab_operator.py:
```diff
@@ -0,0 +1,324 @@
+
+
+def load_image(name, n, m=None, gpu=None, square=None):
+    # Function to load images with certain size
```
I wasn't clear enough. The comments should go into the function body *as docstrings*

arsenal9971 commented:
Done, I thought that would be not necessary since this functions are not necessary going out of the API.

kohr-h commented:
> Done, I thought that would be not necessary since this functions are not necessary going out of the API.

Just think of it as a nice move towards other developers :wink:

kohr-h commented:
Okay, I'll merge this now.

arsenal9971 commented:
Thanks!

kohr-h commented:
> Thanks!

Thank YOU for this great contribution!

adler-j commented:
Enormous thank you @arsenal9971, great work!

arsenal9971 commented:
I hope this is just the first contribution, I have something cooking with the CT reconstruction
